src/libraries/Common/tests/System/Net/Http/Http2LoopbackConnection.cs

```diff
@@ -28,19 +28,68 @@ public class Http2LoopbackConnection : GenericLoopbackConnection
         private readonly TimeSpan _timeout;
         private int _lastStreamId;
         private bool _expectClientDisconnect;
+        private readonly SemaphoreSlim? _readLock;
+        private readonly SemaphoreSlim? _writeLock;
 
         private readonly byte[] _prefix = new byte[24];
         public string PrefixString => Encoding.UTF8.GetString(_prefix, 0, _prefix.Length);
         public bool IsInvalid => _connectionSocket == null;
         public Stream Stream => _connectionStream;
         public Task<bool> SettingAckWaiter => _ignoredSettingsAckPromise?.Task;
 
-        private Http2LoopbackConnection(SocketWrapper socket, Stream stream, TimeSpan timeout, bool transparentPingResponse)
+        private Http2LoopbackConnection(SocketWrapper socket, Stream stream, TimeSpan timeout, Http2Options httpOptions)
         {
             _connectionSocket = socket;
             _connectionStream = stream;
             _timeout = timeout;
-            _transparentPingResponse = transparentPingResponse;
+            _transparentPingResponse = httpOptions.EnableTransparentPingResponse;
+
+            if (httpOptions.EnsureThreadSafeIO)
+            {
+                _readLock = new SemaphoreSlim(1, 1);
+                _writeLock = new SemaphoreSlim(1, 1);
+                _connectionStream = CreateConcurrentConnectionStream(stream, _readLock, _writeLock);
+            }
+
+            static Stream CreateConcurrentConnectionStream(Stream stream, SemaphoreSlim readLock, SemaphoreSlim writeLock)
+            {
+                return new DelegateStream(
+                    canReadFunc: () => true,
+                    canWriteFunc: () => true,
+                    readAsyncFunc: async (buffer, offset, count, cancellationToken) =>
+                    {
+                        await readLock.WaitAsync(cancellationToken);
+                        try
+                        {
+                            return await stream.ReadAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
```

Comment:
Should the `DelegateStream` funcs use `ConfigureAwait` ? Won't it then escape the xUnit synchronization context?